### PR TITLE
Updating command reference

### PR DESCRIPTION
**Pull request by bmmpxf**
Added missing commands, updated existing command descriptions, and unified structure of listing.

### CODE REVIEW

File: public/docs/advanced/commands.md [hunk 1]
```diff
@@ -1,129 +1,215 @@
...
 
-```none
-$ truffle [command] [options]
+All commands are in the following form:
+
+```shell
+truffle [command] [options]
 ```
 
+No arguments given is equivalent to `truffle help`, which will display a list of all commands and then exit.
```
Comment (tcoulter) [inline public/docs/advanced/commands.md]: I got confused on this line (tried to type my confusion but figured it wasn't worth the long stream of consciousness). I'd reword this to say 
> Passing no arguments is equivalent to `truffle help`

or

> If no arguments are given it's equivalent to typing `truffle help`. 
etc.

File: public/docs/advanced/commands.md [hunk 2]
```diff
@@ -1,129 +1,215 @@
...
+Execute build pipeline (if configuration present).
 
-```none
-$ truffle build
+```shell
+truffle build
 ```
 
-See the [Building your application](/docs/getting_started/build) section for more details.
+Requires the `build` key to be present in the configuration. See the [Building your application](/docs/getting_started/build) and [Build processes](/docs/advanced/build_processes) sections for more details.
```
Comment (tcoulter) [inline public/docs/advanced/commands.md]: Not that this is in the scope of work, but I'd add a deprecating warning to this command. e.g., 

> This command is deprecated. Please use third-party build tools like webpack or grunt, or see our Truffle Boxes[LINKME] for an example.

File: public/docs/advanced/commands.md [hunk 3]
```diff
@@ -1,129 +1,215 @@
...
 Run a console with contract abstractions and commands available.
 
-```none
-$ truffle console
+```shell
+truffle console [--network <name>|--verbose-rpc]
 ```
 
-Once the console starts, you can then use your contracts via the command line like you would in your code. Additionally, all truffle commands listed here are available within the console.
+Spawns an interface to interact with contracts via the command line. Additionally, many Truffle commands are additionally available within the console (without the `truffle` prefix).
```
Comment (tcoulter) [inline public/docs/advanced/commands.md]: The word `additionally` is used twice in this sentence. One of them is likely unnecessary.

File: public/docs/advanced/commands.md [hunk 4]
```diff
@@ -1,129 +1,215 @@
...
+
+Requires an external RPC, such as [Ganache](/docs/ganache/using) or geth. For the console that generates its own RPC, use `truffle develop`.
```
Comment (tcoulter) [inline public/docs/advanced/commands.md]: Instead RPC, use 
> Ethereum client

Comment (tcoulter) [reply public/docs/advanced/commands.md]: And for the second instance of RPC, use
> For a console that creates a development and test environment for you, use `truffle develop`.

File: public/docs/advanced/commands.md [hunk 5]
```diff
@@ -1,129 +1,215 @@
...
+
+See the [Using the console](/docs/getting_started/console) section for more details.
 
 Optional parameters:
 
-* `--network name`: Specify the network to use.
+* `--network <name>`: Specify the network to use. Network name must exist in the configuration.
 * `--verbose-rpc`: Log communication between Truffle and the RPC.
```
Comment (tcoulter) [inline public/docs/advanced/commands.md]: Should probably change RPC here too to something fancy.

File: public/docs/advanced/commands.md [hunk 6]
```diff
@@ -1,129 +1,215 @@
...
-### compile
+### create
 
-Intelligently compile your contracts. This will only compile contracts that have changed since the last compile, unless otherwise specified.
+Helper to create new contracts, migrations and tests.
 
-```none
-$ truffle compile
+```shell
+truffle create (contract|migration|test) <ArtifactName>
```
Comment (tcoulter) [inline public/docs/advanced/commands.md]: Nice merge of the `create` command. 👍

File: public/docs/advanced/commands.md [hunk 7]
```diff
@@ -1,129 +1,215 @@
...
-Helper method to scaffold a new test for a contract. Name must be camel-case.
+Open a console with a local TestRPC
 
-```none
-$ truffle create test MyTest
+```shell
+truffle develop
 ```
 
+Spawns a local development blockchain, and allows you to interact with contracts via the command line. Additionally, many Truffle commands are additionally available within the console (without the `truffle` prefix).
```
Comment (tcoulter) [inline public/docs/advanced/commands.md]: Double `additionally` here too.

File: public/docs/advanced/commands.md [hunk 8]
```diff
@@ -1,129 +1,215 @@
...
+
+If you want an interactive console but want to use an existing blockchain, use `truffle console`.
```
Comment (tcoulter) [inline public/docs/advanced/commands.md]: Love it.

File: public/docs/advanced/commands.md [hunk 9]
```diff
@@ -1,129 +1,215 @@
...
+
+See the [Using the console](/docs/getting_started/console) section for more details.
+
+
 ### exec
 
-Execute a Javascript file within the Truffle environment. This will include `web3`, set the default provider based on the network specified (if any), and include your contracts as global objects while executing the script. Your script must export a function that Truffle can run. See the [Writing external scripts](/docs/getting_started/scripts) section for more details.
+Execute a JS module within this Truffle environment.
```
Comment (tcoulter) [inline public/docs/advanced/commands.md]: To me, `this` here is confusing. Which environment is dependent on the network specified (if specified). I'd just us `a`

File: public/docs/advanced/commands.md [hunk 10]
```diff
@@ -1,129 +1,215 @@
...
-Create a completely new app within the current working directory. Will add default contracts, tests and frontend configuration.
+Initialize new Ethereum project with example contracts and tests.
 
-```none
-$ truffle init
+```shell
+truffle init
 ```
 
+Creates a new and empty Truffle project within the current working directory. Will add default contracts, tests and frontend configuration. Takes no arguments.
```
Comment (tcoulter) [inline public/docs/advanced/commands.md]: This no longer creates default contracts, tests and frontend configuration. So that should likely be removed. Also this sentence doesn't match the `empty Truffle project` mentioned in the previous sentence. 

This leaves the sentence: 
> Creates a new and empty Truffle project within the current working directory

This is a good summary sentence, but you added another above. I'd choose one. (I like the one on this line, your call).

File: public/docs/advanced/commands.md [hunk 11]
```diff
@@ -1,129 +1,215 @@
...
 Optional parameters:
 
 * `--reset`: Run all migrations from the beginning, instead of running from the last completed migration.
-* `-f number`: Run contracts from a specific migration.
-* `--network name`: Specify the network to use, saving artifacts specific to that network.
-* `--compile-all`: Compile all contracts instead of intelligently choosing.
+* `-f <number>`: Run contracts from a specific migration. The number refers to the prefix of the migration file.
+* `--network <name>`: Specify the network to use, saving artifacts specific to that network. Network name must exist in the configuration.
+* `--compile-all`: Compile all contracts instead of intelligently choosing which contracts need to be compiled.
 * `--verbose-rpc`: Log communication between Truffle and the RPC.
```
Comment (tcoulter) [inline public/docs/advanced/commands.md]: RPC is wrong here as well (this stuff is way old). I'd say `the Ethereum client` here.

File: public/docs/advanced/commands.md [hunk 12]
```diff
@@ -132,55 +218,94 @@ Optional parameters:
...
+truffle serve [-p <port>] [--network <name>]
 ```
 
 Optional parameters:
 
-* `-p port`: Specify the port to serve on. Default is 8080.
-* `--network name`: Specify the network to use, using artifacts specific to that network.
+* `-p <port>`: Specify the port to serve on. Default is 8080.
+* `--network <name>`: Specify the network to use, using artifacts specific to that network. Network name must exist in the configuration.
+
```
Comment (tcoulter) [inline public/docs/advanced/commands.md]: Like `build` above, I'd add a statement to the end of this one saying that it's deprecated. Something like:

> This command is deprecated. Please use third-party development servers like `webpack-dev-server` instead. See our Webpack Truffle box for an example.

If you can link the above sentence that'd be awesome too!

File: public/docs/advanced/commands.md [hunk 13]
```diff
@@ -132,55 +218,94 @@ Optional parameters:
...
 
 ### test
 
-Run all tests within the `./test` directory, or optionally run a single test.
+Run Mocha and Solidity tests.
```
Comment (tcoulter) [inline public/docs/advanced/commands.md]: I'd replace Mocha with "Javascript" here, only because that's how we differentiate in other places. We should mention Mocha, but I believe that's listed on the Testing pages. Probably best to link to those pages somewhere under this command reference like you did with other ones.

File: public/docs/advanced/commands.md [hunk 14]
```diff
@@ -132,55 +218,94 @@ Optional parameters:
...
 
-* `--compile-all`: Compile all contracts instead of intelligently choosing.
+* `--compile-all`: Compile all contracts instead of intelligently choosing which contracts need to be compiled.
+* `--network <name>`: Specify the network to use, using artifacts specific to that network. Network name must exist in the configuration.
 * `--verbose-rpc`: Log communication between Truffle and the RPC.
-* `--network name`: Specify the network to use, using artifacts specific to that network.
+
+### unbox
+
+Unbox Truffle project.
```
Comment (tcoulter) [inline public/docs/advanced/commands.md]: Probably need to fill this sentence out. `Unbox a pre-built Truffle project.`

File: public/docs/advanced/commands.md [hunk 15]
```diff
@@ -132,55 +218,94 @@ Optional parameters:
...
+
+```shell
+truffle unbox box_name
```
Comment (tcoulter) [inline public/docs/advanced/commands.md]: This syntax seems different from your others. Shouldn't it be the following?
 > `truffle unbox <box name>`

or just 

>` truffle unbox <name>`

File: public/docs/advanced/commands.md [hunk 16]
```diff
@@ -132,55 +218,94 @@ Optional parameters:
...
-Watch for changes to contracts, app and configuration files. When there's a change, rebuild the app if necessary.
+Watch filesystem for changes and rebuild the project automatically.
 
-```none
-$ truffle watch
+```shell
+truffle watch
 ```
+
+This command will initiate a watch for changes to contracts, application, and configuration files. When there's a change, it will rebuild the app as necessary. Similar to `truffle serve`, but without the web server component.
```
Comment (tcoulter) [inline public/docs/advanced/commands.md]: This is another command that should be deprecated. For this one, something like:

> This command is deprecated. Please use external tools to watch for filesystem changes and rerun tests.

File: public/docs/advanced/commands.md (hunk 2 above)
Comment (bmmpxf) [reply public/docs/advanced/commands.md]: That's good to know. I wasn't sure which of these commands were ongoing concerns.

File: public/docs/advanced/commands.md (hunk 5 above)
Comment (bmmpxf) [reply public/docs/advanced/commands.md]: Might as well use "Ethereum client", as that's what we subbed for RPC above.

The command name itself (`--verbose-rpc`) might also be worth revisiting.

File: public/docs/advanced/commands.md (hunk 9 above)
Comment (bmmpxf) [reply public/docs/advanced/commands.md]: Not sure where I got "this" from. I like how it originally read now.

File: public/docs/advanced/commands.md (hunk 10 above)
Comment (bmmpxf) [reply public/docs/advanced/commands.md]: The top line of all of these entries match the blurb that shows up on the `help` output.  But good point about what `init` does. I updated the blurb in the truffle-core project (and here).

File: public/docs/advanced/commands.md (hunk 13 above)
Comment (bmmpxf) [reply public/docs/advanced/commands.md]: Sure. That line is what's listed in `truffle help`, so I went ahead and changed it there (and here) to "Run JavaScript and Solidity tests"

File: public/docs/advanced/commands.md (hunk 14 above)
Comment (bmmpxf) [reply public/docs/advanced/commands.md]: Will edit the description in `truffle-core`.

So as to avoid using the word in its definition, I think "Downloads a Truffle Box, a pre-built Truffle project" might be more appropriate.

File: public/docs/advanced/commands.md (hunk 15 above)
Comment (bmmpxf) [reply public/docs/advanced/commands.md]: Right!

File: public/docs/advanced/commands.md (hunk 16 above)
Comment (bmmpxf) [reply public/docs/advanced/commands.md]: Got it.

Should we be notating these deprecations somewhere else? Perhaps in the command description in `truffle help` or as a console note during their execution?

**Comment by bmmpxf** [comment]
@tcoulter Thanks for the review! I think this is a really valuable page, and now it's definitely more rock solid. (Plus, as a bonus, we have updated command descriptions in the Truffle app itself.)
